### PR TITLE
Don't force set health status if there is an initial snapshot

### PR DESCRIPTION
**Pull request by adamint**
## Description

There was one regression and one red herring from the health check playground project.

Regression: we should not force set the health status to healthy if there is an initial snapshot.

Red herring: if a health check is not run, the health status is not recomputed. Because of this, we should be specifying the health status in the initial snapshot in the playground project resources. 

@JamesNK @drewnoakes 

## Checklist

- Is this feature complete?
  - [X] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [X] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [X] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [X] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [X] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/6367)

### CODE REVIEW

File: tests/Aspire.Hosting.Tests/Dashboard/DashboardServiceTests.cs [hunk 1]
```diff
@@ -87,20 +87,14 @@ public async Task WatchResources_ResourceHasCommands_CommandsSentWithResponse()
...
             updateState: c => ApplicationModel.ResourceCommandState.Enabled,
             displayDescription: "Display description!",
-            parameter: new [] {"One", "Two"},
+            parameter: new[] { "One", "Two" },
             confirmationMessage: "Confirmation message!",
             iconName: "Icon name!",
             iconVariant: ApplicationModel.IconVariant.Filled,
             isHighlighted: true);
 
-        await resourceNotificationService.PublishUpdateAsync(testResource, s =>
```
Comment (adamint) [inline tests/Aspire.Hosting.Tests/Dashboard/DashboardServiceTests.cs:96]: diff is messed up, look at side by side

File: src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs [hunk 2]
```diff
@@ -395,8 +395,17 @@ public Task PublishUpdateAsync(IResource resource, string resourceId, Func<Custo
...
-        // A resource is also healthy if it has no health check annotations and is in the running state.
-        if (previousState.HealthStatus is not HealthStatus.Healthy && !resource.TryGetAnnotationsIncludingAncestorsOfType<HealthCheckAnnotation>(out _) && previousState.State?.Text == KnownResourceStates.Running)
+        var hasHealthChecks = resource.TryGetAnnotationsIncludingAncestorsOfType<HealthCheckAnnotation>(out _);
+        var hasInitialSnapshot = resource.TryGetLastAnnotation<ResourceSnapshotAnnotation>(out _);
+
+        // A resource is also healthy if it
+        // - has no health check annotations
+        // - is in the running state
+        // - was not started with an initial snapshot
+        if (previousState.HealthStatus is not HealthStatus.Healthy
```
Comment (JamesNK) [inline src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs:405]: If there is no initial snapshot, and there is no annotation, doesn't that mean `HealthStatus` must be null? That seems more specific test than not healthy.

File: src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs [hunk 3]
```diff
@@ -395,8 +395,17 @@ public Task PublishUpdateAsync(IResource resource, string resourceId, Func<Custo
...
+            && !hasHealthChecks
+            && previousState.State?.Text == KnownResourceStates.Running
```
Comment (JamesNK) [inline src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs:407]: Put the previousState checks together? Then inside the if statement, resolve the attributes then. That means they're only resolved when needed.

File: src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs (hunk 2 above)
Comment (adamint) [reply src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs]: > HealthStatus must be null

Yeah, good point. There's nothing that could have affected the health status. Updated

File: tests/Aspire.Hosting.Tests/Dashboard/DashboardServiceTests.cs (hunk 1 above)
Comment (JamesNK) [reply tests/Aspire.Hosting.Tests/Dashboard/DashboardServiceTests.cs:96]: What do you mean? Is this just whitespace change? Note that it will conflict with a change I've just made here: https://github.com/dotnet/aspire/pull/6366

File: src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs (hunk 3 above)
Comment (adamint) [reply src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs]: done, thanks

File: tests/Aspire.Hosting.Tests/Dashboard/DashboardServiceTests.cs [hunk 4]
```diff
@@ -130,6 +124,46 @@ await resourceNotificationService.WaitForResourceAsync(testResource.Name, r =>
...
+
+        // Act
+        var task = dashboardService.WatchResources(
+            new WatchResourcesRequest(),
+            writer,
+            context);
+
+        // Assert
+        var initialRead = await writer.ReadNextAsync();
+        var initialSnapshot = Assert.Single(initialRead.InitialData.Resources);
```
Comment (JamesNK) [inline tests/Aspire.Hosting.Tests/Dashboard/DashboardServiceTests.cs:160]: This will be flaky. There is no guarentee that published resource will be in the initial data returned by the service.

Waiting for the resource mostly fixes it (see above test). However I believe you need to wait on another type to properly fix it. See https://github.com/dotnet/aspire/pull/6366 and the changes I made on the `DashboardDataService`.

File: src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs [hunk 5]
```diff
@@ -395,10 +395,22 @@ public Task PublishUpdateAsync(IResource resource, string resourceId, Func<Custo
...
+        // - is in the running state
+        // - was not started with an initial snapshot
+        if (previousState.HealthStatus is null)
         {
-            return previousState with { HealthStatus = HealthStatus.Healthy };
+            var hasHealthChecks = resource.TryGetAnnotationsIncludingAncestorsOfType<HealthCheckAnnotation>(out _);
+            var hasInitialSnapshot = resource.TryGetLastAnnotation<ResourceSnapshotAnnotation>(out _);
+
+            if (!hasHealthChecks
+                && previousState.State?.Text == KnownResourceStates.Running
```
Comment (JamesNK) [inline src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs:409]: nit: Check at the same time as health status. It is cheaper that getting annotations.

File: src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs [hunk 6]
```diff
@@ -395,10 +395,22 @@ public Task PublishUpdateAsync(IResource resource, string resourceId, Func<Custo
...
+        // A resource is also healthy if it
+        // - has a null health status (wasn't set by a health check nor initial snapshot)
+        // - has no health check annotations
+        // - is in the running state
+        // - was not started with an initial snapshot
+        if (previousState.HealthStatus is null)
         {
-            return previousState with { HealthStatus = HealthStatus.Healthy };
+            var hasHealthChecks = resource.TryGetAnnotationsIncludingAncestorsOfType<HealthCheckAnnotation>(out _);
+            var hasInitialSnapshot = resource.TryGetLastAnnotation<ResourceSnapshotAnnotation>(out _);
```
Comment (JamesNK) [inline src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs:406]: I don't think this works. Someone could add a health report with ResourceNotificationService after starting a resource.

Comment (JamesNK) [reply src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs:406]: For example:

```cs
await notificationService.PublishUpdateAsync(
    resource,
    state => state with
    {
        State = new("Running", "success"),
        HealthReports = [new HealthReportSnapshot($"{name}_check", status, description, exception)]
    });
```

File: tests/Aspire.Hosting.Tests/Dashboard/DashboardServiceTests.cs (hunk 4 above)
Comment (adamint) [reply tests/Aspire.Hosting.Tests/Dashboard/DashboardServiceTests.cs:160]: so is this dependent on #6366?

Comment (JamesNK) [reply tests/Aspire.Hosting.Tests/Dashboard/DashboardServiceTests.cs:160]: You can do what the test above is currently doing. It's 99.9% stable. Create an issue to update it once 6366 is merged.

File: tests/Aspire.Hosting.Tests/Dashboard/DashboardServiceTests.cs (hunk 1 above)
Comment (adamint) [reply tests/Aspire.Hosting.Tests/Dashboard/DashboardServiceTests.cs:96]: No, a new test has been added in that placement - we can fix the merge conflict

File: src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs (hunk 6 above)
Comment (adamint) [reply src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs]: What do you think about 
- removing this method entirely
- on the dashboard side, say a resource with a null health status is healthy if it is running with 0 health reports
- not make any assumptions other than above about the resource status - if users are updating health reports themselves, they need to handle health status too

@JamesNK

Comment (JamesNK) [reply src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs]: Here is what you could do:

1. Exit out of the method if `HealthCheckAnnotation` is set. Assume that the system will manage updating health status
2. If state is not running then set HealthStatus to null. If it's not running then there is no status. This is needed for resources that start/stop. Shouldn't have a healthy status if the app isn't running.
3. If state is running then look at health reports to figure out status. No reports = healthy. A degragded report = degradged, etc.

In other words, let aspire manage the health status if `HealthCheckAnnotation` is set, otherwise the health reports on the snapshot are the source of truth.

Having both health status and health reports on the snapshot be mutable is a point of friction. Either they can go out of sync (could be unexpected) or the health reports overwrite what is manually set.
